### PR TITLE
#1917 validation prevent mobile page loading

### LINKED ISSUE #1917: Prevent laptop users from going to mobile validation page
It doesn't look great
mobile-validate-laptop

### PR DESCRIPTION
**Pull request by kumararoosh**
Modified ValidationController.scala by adding a function that determines if the user is on a mobile device. A regular expression is used to examine the User-Agent. There is no longer a redirect after loading the mobile page, instead laptop users are immediately served the audit page.

Resolves #1917

### CODE REVIEW

File: app/controllers/ValidationController.scala [hunk 1]
```diff
@@ -30,6 +32,20 @@ class ValidationController @Inject() (implicit val env: Environment[User, Sessio
   val validationMissionStr: String = "validation"
   val mobileValidationMissionStr: String = "validation"
   val rapidValidationMissionStr: String = "rapidValidation"
+  val mobileOS: Regex = "(iPhone|webOS|iPod|Android|BlackBerry|mobile|SAMSUNG|IEMobile|OperaMobi)".r
```
Comment (misaugstad) [inline app/controllers/ValidationController.scala]: Where did you find this particular list for the regex? I feel like this probably is not comprehensive. This stackoverflow post seems to suggest using a more comprehensive regex: https://stackoverflow.com/questions/11381673/detecting-a-mobile-browser

And I know we've been saying mobile, but we want to include tablets with mobile devices. Our website doesn't support touchscreens well.

Comment (misaugstad) [inline app/controllers/ValidationController.scala]: This variable is only used in the function that follows it, might as well just define the variable in that function.

File: app/controllers/ValidationController.scala [hunk 2]
```diff
@@ -57,11 +73,11 @@ class ValidationController @Inject() (implicit val env: Environment[User, Sessio
     */
   def mobileValidate = UserAwareAction.async { implicit request =>
     val ipAddress: String = request.remoteAddress
-
+    println(isMobile(request))
```
Comment (misaugstad) [inline app/controllers/ValidationController.scala]: Make sure to remove this print statement.

**Comment by kumararoosh** [comment]
Made the changes requested. Unfortunately, there is no mobile detection regex that already exists for Scala. I was able to write my own regex that had most of the mobile devices from javascript, but some of the more complex regex terms couldn't be translated.

**Comment by misaugstad** [comment]
do you know what in particular is missing from the regex then?

**Comment by kumararoosh** [comment]
All that is missing are some more obscure tablets. There is a list of user-agents at https://developers.whatismybrowser.com/useragents/explore/operating_system_name/ but the listing is disabled unfortunately

**Comment by misaugstad** [comment]
@kumararoosh were you able to access the website using your phone? I haven't been able to access it myself. It looks like the `isMobile` function is returning false for my iPhone. I printed out what the `agent` variable in the `isMobile` function looks like and got this:

```
Mozilla/5.0 (iPhone; CPU iPhone OS 13_3_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.5 Mobile/15E148 Safari/604.1
```

**Comment by kumararoosh** [comment]
@misaugstad it appears that I implemented a flawed Regex. I will reconstruct the regex from scratch to hopefully get a fully working regex

**Comment by kumararoosh** [comment]
The final regex is now (iPhone|webOS|iPod|Android|BlackBerry|mobile|SAMSUNG|IEMobile|OperaMobi|BB10|iPad|Tablet)

This includes iPhones, iPads, iPods, all Android phones, all BlackBerrys, and all Windows phones. I tested these cases by using a User-Agent manipulator tool in Chrome, so it should work with all phones unless the user manually modifies their own User-Agent

**Comment by kumararoosh** [comment]
All the small edits are finished, and the branch should be able to merge

File: public/javascripts/SVValidate/mobile/mobileValidate.js [hunk 3]
```diff
@@ -1,4 +1,4 @@
-$(document).ready(function() {
+$(document).ready(function() {
```
Comment (misaugstad) [inline public/javascripts/SVValidate/mobile/mobileValidate.js]: Oh man, so close :joy: Looks like you accidentally also added a few extra spaces to the end of this line when you were removing that other line that was accidentally added. Can you fix this too?

Just make sure to run `git diff` on each file before you add and commit them to make sure there are only changes listed there that you intended to make!

**Comment by kumararoosh** [comment]
I believe the all the changes should be done

**Pull request merged**